test/e2e/network/network_policy.go

```diff
@@ -352,12 +486,552 @@ var _ = SIGDescribe("NetworkPolicy", func() {
+			podServerStatus, err := f.ClientSet.CoreV1().Pods(f.Namespace.Name).Get(podServer.Name, metav1.GetOptions{})
+			if err != nil {
+				Expect(err).NotTo(HaveOccurred())
+			}
+
+			podServerCIDR := fmt.Sprintf("%s/32", podServerStatus.Status.PodIP)
+
+			// Creating pod-b and service-b
+			podServerB, serviceB = createServerPodAndService(f, f.Namespace, "pod-b", []int{80})
+			By("Waiting for pod-b to be ready", func() {
+				err = f.WaitForPodReady(podServerB.Name)
+				Expect(err).NotTo(HaveOccurred())
+			})
+			defer cleanupServerPodAndService(f, podServerB, serviceB)
+
+			// Wait for podServerB with serviceB to be ready
+			err = framework.WaitForPodRunningInNamespace(f.ClientSet, podServerB)
+			Expect(err).NotTo(HaveOccurred())
+
+			// Before creation of the policy client-a should be able to connect to server-b
+			By("Creating client-a which should not be able to contact the server-b.", func() {
+				testCanConnect(f, f.Namespace, "client-a", serviceB, 80)
+			})
+
+			policy := &networkingv1.NetworkPolicy{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: f.Namespace.Name,
+					Name:      "allow-client-a-via-cidr-egress-rule",
+				},
+				Spec: networkingv1.NetworkPolicySpec{
+					// Apply this policy to the Server
+					PodSelector: metav1.LabelSelector{
+						MatchLabels: map[string]string{
+							"pod-base-name": "client-a",
+						},
+					},
+					PolicyTypes: []networkingv1.PolicyType{networkingv1.PolicyTypeEgress},
+					// Allow traffic to only one CIDR block.
+					Egress: []networkingv1.NetworkPolicyEgressRule{{
+						To: []networkingv1.NetworkPolicyPeer{
+							{
+								IPBlock: &networkingv1.IPBlock{
+									CIDR: podServerCIDR,
+								},
+							},
+							// Need to allow traffic to DNS for name relsolution
+							{
+								NamespaceSelector: &metav1.LabelSelector{
+									MatchLabels: map[string]string{},
+								},
+								PodSelector: &metav1.LabelSelector{
+									MatchLabels: map[string]string{
+										"k8s-app": "kube-dns",
+									},
+								},
+							},
+						},
+					}},
+				},
+			}
+
+			policy, err = f.ClientSet.NetworkingV1().NetworkPolicies(f.Namespace.Name).Create(policy)
+			Expect(err).NotTo(HaveOccurred())
+			defer cleanupNetworkPolicy(f, policy)
+
+			By("Creating client-a which should not be able to contact the server-b.", func() {
+				testCannotConnect(f, f.Namespace, "client-a", serviceB, 80)
+			})
+
+			By("Creating client-a which should be able to contact the server.", func() {
+				testCanConnect(f, f.Namespace, "client-a", service, 80)
+			})
+		})
+	})
+
+	Context("NetworkPolicy between pods, each containing a server container and client container", func() {
+		BeforeEach(func() {
+			By("Creating a pod with simple server container, that serves on port 80 and 81, and a client container.")
+			podServer, service = createPodWithClientServerAndService(f, f.Namespace, "pod-a", []int{80, 81}, "client-a", "pod-a-server-responded")
```

Comment:
you don't need to do that; you can just create a client and a server with the same labels, and you'd be testing the same thing

Reply:
I have tried the way you suggest in the comment. I have written this test in this way for two reasons:
1. I created a client and a server with different labels for one test, then deleted them and recreate another set of server and client, with labels exchanged. In this case, I found that there are times when the creation of the second set of server and client starts before the first set is deleted completely. This leads to failure of the test as object being created already exists. This behavior is flaky. Prior to my changes, pod name and pod label were same. So I have made the pod name distinct by adding the random string (this is also the answer for the question you asked earlier).

2. For testing of this scenario, it seemed better to test without deleting the pods.

Please let me know your thoughts.

Reply:
@danwinship  I would suggest to use exact same pod for connectivity check before and after applying network policy. If we recreate another pair of client/server pods with same label, those pods may be rescheduled onto different nodes and the connectivity between them could be affected by other aspect instead of network policy.